README

WebSockets for NaviServer
=========================
Release 0.3
-----------

    neumann@wu-wien.ac.at
    wolfgang.winkler@digital-concepts.com

This is a NaviServer module that implements WebSockets for NaviServer
and provides an API for it. The WebSockets are integrated in
NaviServer using the ns_chan command. The current implementation
requires NaviServer 4.99.20 or newer.

Connections are established via ws::handshake, which is typically
initiated via JavaScript. Common libraries such as jQueries provided
support for WebSockets.

The WebSocket protocol (RFC 6455) defines a ws:// and wss:// prefix to
indicate a WebSocket or a WebSocket Secure connection. Both schemes
use the HTTP upgrade mechanism to upgrade to the WebSocket protocol.
For WebSocket Secure connection, the NaviServer module nsssl has to be
installed and configured.  For incoming HTTP/HTTPS connections the
upgrade is provided via the command ws::handshake, which returns a
handle for the connection channel. This handle has to be used for all
other operations on the WebSocket, such as sending data to a single
channel, or for subscribing / unsubscribing the channel from a
subscription. The API provides means for encoding and decoding
websocket messages and to send messages as single- or multi-casts
using a subscriber interface.x

The package contains two sample WebSocket applications:

* chat: A simple chat application with the mere purpose to provide an
  example how to use WebSockets
  
* log-view: A WebSocket application that allows live viewing of log
  contents. It uses on a tcl-based "tail -f" like command that watches
  certain files (e.g. access.log, or error.log) and reports recent
  changes to the subscribed clients.

By default, the sample applications are deactivated. See below in the
configuration section how to activate the sample applications.

The implementation is based on nsf, which is available from e.g.
http://next-scripting.org/
    
***

Configuration:
--------------

In order to configure WebSockets, add the following lines to the
config file of NaviServer. The section module/websocket/$appname is
optional and allows one to configure the named websocket applications.

When the section module/websocket/$appname is missing, or the "urls"
parameter is empty, then the application is deactivated.  Via the urls
parameters, multiple urls can be specified, under which the WebSocket
applications are available. This way, one can e.g. define multiple
chats with different subscriber lists.

Be aware that the log-viewer might reveal internal information, so be
careful when registering it on public sites. In the example below, we
provide it under "/admin/", which is restricted for OpenACS to site
wide administrators.

    ns_section "ns/server/${server}/modules" {
       ns_param websocket tcl
    }
    ns_section "ns/server/${server}/module/websocket/chat" {
       ns_param urls     /websocket/chat
    }
    ns_section "ns/server/${server}/module/websocket/log-view" {
       ns_param urls     /admin/websocket/log-view
       ns_param refresh  1000   ;# refresh time for file watcher in milliseconds
    }


Installation:
-------------

    make install


Usage:
------

The provided WebSocket server API consists of the following commands:

    ws::handshake ?-readevent /boolean/? ?-callback /script/?
    ws::log /msg/
    ws::multicast ?-exclude /value/? /subscription/ /msg/
    ws::io /channel/ /callback/ /when/
    ws::send /channel/ /msg/
    ws::subscribe /channel/ /subscription/
    ws::unsubscribe /channel/ /subscription/

The provided WebSocket client API consists of the following commands:

    ws::client::open /websocketurl/
    ws::client::send /channel/ /textmessage/
    ws::client::receive /channel/
    ws::client::close /channel/

See the included example WebSocket server applications for usage patterns.

Authors:
--------

    Gustaf Neumann neumann@wu-wien.ac.at
    Wolfgang Winkler wolfgang.winkler@digital-concepts.com




Changelog
commit b30472fb246de15e098fdfa3cc112e47c28d88d0
Author: Gustaf Neumann <neumann@wu-wien.ac.at>
Date:   2021-07-26

    guard "dict get" operation by "dict exists" statement

M	websocket-procs.tcl

commit d72d6c5ae57d4804ae599b4dd274f9905833d56a
Author: Gustaf Neumann <neumann@wu-wien.ac.at>
Date:   2021-06-21

    fix typo

M	websocket-procs.tcl

commit 373c4ecf12b446f9b063e708abd3553058dd2223
Author: Gustaf Neumann <neumann@wu-wien.ac.at>
Date:   2021-01-04

    added deprecated function ::ws::build_msg for providing better backward compatibility

M	websocket-procs.tcl

commit 474267bfbbdbff78a9288ef35289b1ba63c9062a
Author: Gustaf Neumann <neumann@wu-wien.ac.at>
Date:   2021-01-01

    use ns_connchan "wsencode" instead of "wsframe" since it leaves room for a decode operation

M	websocket-procs.tcl

commit d7e5d364083a62e8c6ce07ab4ac4832f7d0b4a33
Author: Gustaf Neumann <neumann@wu-wien.ac.at>
Date:   2020-12-31

    Major rewrite, using C-level implementing in ns_connchan
    
    This version make uses of the new C-level support for WebSockets.
    The new version is several orders of magnitudes faster than the
    previous version, it is more robust (C-level handling of
    partial read and write operations) and it is more complete
    than previous versions (handling segmented messages)

M	README
M	websocket-procs.tcl

commit bd61f0f932e83c8cea9da6163f4511c7b36bfdc9
Author: Gustaf Neumann <neumann@wu-wien.ac.at>
Date:   2020-12-16

    make parsing slightly more robust, use atomic nsv updates

M	websocket-procs.tcl

commit 0edbf1bdabc9b2ac33d6d6f72003500f50a2585d
Author: Gustaf Neumann <neumann@wu-wien.ac.at>
Date:   2020-11-27

    provide name for mutex lock to ease tracing of long locks

M	websocket-procs.tcl

commit ab80f28a3c91675c242dabb49ca460e24364f65d
Author: Gustaf Neumann <neumann@wu-wien.ac.at>
Date:   2020-10-21

    simplify code, and don't mask beyond end of payload

M	websocket-procs.tcl

commit 56a76a538425dc26fc030f6a62956b46177f5494
Author: Gustaf Neumann <neumann@wu-wien.ac.at>
Date:   2020-10-21

    handling rest bytes (package longer than a WS frame)

M	websocket-procs.tcl

commit 82800461b1403eda6113cd7f41f55aae39dc4b07
Author: Gustaf Neumann <neumann@wu-wien.ac.at>
Date:   2020-07-15

    fixed 2 potential issues:
    
    a) in case the websocket client sends messages >16KB, the packet is split (at least with HTTPS), which was not handled correctly before.
    b) the code was converting always to UTF8, also in binary cases

M	websocket-procs.tcl

commit a7db0fb1383ffd52731dd4d9b616698814abf213
Author: Gustaf Neumann <neumann@wu-wien.ac.at>
Date:   2020-07-14

    add UTF-8 decoding of the returned text message

M	websocket-procs.tcl

commit feaa885284f08e9e80664fedf6c6d4539af242ee
Author: Gustaf Neumann <neumann@wu-wien.ac.at>
Date:   2020-07-14

    Added client-side SNI (requires recent version of NaviServer)

M	websocket-procs.tcl

commit 521c9bfa70ffb4c1730225e8f18f24697c40de5d
Author: Gustaf Neumann <neumann@wu-wien.ac.at>
Date:   2020-07-13

    bump version number

M	README

commit 07068de20fc2ec48448ff84624bec4e4526ae58f
Author: Gustaf Neumann <neumann@wu-wien.ac.at>
Date:   2020-07-13

    added simple websocket client interface

M	README
M	log-view.tcl
M	websocket-procs.tcl

commit 5008634dde2a59a154a16c008c918baac4820a4f
Author: Gustaf Neumann <neumann@wu-wien.ac.at>
Date:   2019-08-10

    improve spelling

M	README

commit b6b7c769651c27322a72feb9c2d31d70bbc3be8b
Author: Gustaf Neumann <neumann@wu-wien.ac.at>
Date:   2019-08-04

    Compatibility for Tcl 8.7, which has no "identity" encoding anymore

M	websocket-procs.tcl

commit 6ca02a329a8cef8a5dedde6b1fafabda0780982f
Author: Gustaf Neumann <neumann@wu-wien.ac.at>
Date:   2019-08-04

    fix startup messages

M	chat.tcl
M	log-view.tcl

commit 22bd4f640c2e9f1180d5f266add51227c46061c6
Author: Gustaf Neumann <neumann@wu-wien.ac.at>
Date:   2018-11-04

    downgrade message from "error" to "warning", since exception during log rotate are not unlikely

M	log-view.tcl

commit 2cc014070851e321a53a658d25ad4a88889e1f63
Author: Gustaf Neumann <neumann@wu-wien.ac.at>
Date:   2018-08-15

    whitespace cleanup and make spelling more consistent

M	chat.tcl
M	log-view.tcl
M	websocket-procs.tcl

commit a3f56089089c769dfb0bc10383d24de087e0bca2
Author: Gustaf Neumann <neumann@wu-wien.ac.at>
Date:   2018-05-05

    Make sure, to send the upgrade header as a single block

M	websocket-procs.tcl

commit 54ed37e3e9bde435604afccba8ea5cd8e89ec4ff
Author: Gustaf Neumann <neumann@wu-wien.ac.at>
Date:   2018-05-05

    remove duplicated word

M	README

commit 48d7984f10302c25807ac7ff244b514ef1898072
Author: Gustaf Neumann <neumann@wu-wien.ac.at>
Date:   2017-04-22

    Fix spelling errors

M	log-view.tcl
M	websocket-procs.tcl

commit 7695e8d4ec115e824ea8f3b321822d54afae3244
Author: Gustaf Neumann <neumann@wu-wien.ac.at>
Date:   2016-01-11

    - make test more precise

M	websocket-procs.tcl

commit 7f6da2e95123b05b5463e8b38ff8c8bea92eeb24
Author: Gustaf Neumann <neumann@wu-wien.ac.at>
Date:   2015-02-23

    - improvided links to cdn in sample scripts
    - updated README file

M	README
M	chat.adp
M	log-view.adp

commit 62a51aced13558c84d97aca963372ed67e023c06
Author: Gustaf Neumann <neumann@wu-wien.ac.at>
Date:   2015-01-05

    - improve README

M	README

commit 1756ceb5fda41a99d09ea3014849c4ed2ef1c4e3
Author: Gustaf Neumann <neumann@wu-wien.ac.at>
Date:   2015-01-05

    - remove debugging line, make output in error.log more consistent

M	log-view.tcl


chat.tcl
#
# Chat demo application for WebSockets.
#
# This demo program shows how to establish a WebSocket connection and
# how to multicast messages. This script is accompanied by chat.adp,
# which resides in the same directory as this script.
#
# Gustaf Neumann, Jan 2015
#
package require nsf

#
# Get configured URLs
#
set URLs [ns_config ns/server/[ns_info server]/module/websocket/chat urls]

if {$URLs eq ""} {
    ns_log notice "WebSocket: no chat configured ([info script])"
    return
}

#
# Register WebSocket chat under every configured URL
#
foreach url $URLs {
    ns_log notice "WebSocket: chat available under $url"
    ns_register_adp  GET $url [file dirname [info script]]/chat.adp
    ns_register_proc GET $url/connect ::ws::chat::connect
}

namespace eval ws::chat {
    #
    # The proc "connect" is called, whenever a new WebSocket is
    # established.  The chat is named via the url to allow multiple
    # independent chats on different URLs.
    #
    nsf::proc connect {} {
        set chat [ns_conn url]
        set channel [ws::handshake -callback [list ws::chat::send_to_all -ip [ns_conn peeraddr] -chat $chat]]
        ws::subscribe $channel $chat
    }

    #
    # Whenever we receive a message, send it to all subscribers of the
    # chat, except the current one.
    #
    nsf::proc send_to_all {{-ip ""} {-chat "chat"} channel msg} {
        if {$ip ne ""} {
            set msg "$ip: $msg"
        }
        ::ws::multicast -exclude [list $channel] $chat [::ws::build_msg $msg]
    }
}

#
# Local variables:
#    mode: tcl
#    tcl-indent-level: 4
#    indent-tabs-mode: nil
# End:


log-view.tcl
#
# Log-view demo application for WebSockets.
#
# This demo program shows how to establish a WebSocket connection and
# how to multicast messages. This script is accompanied by log-view.adp,
# which resides in the same directory as this script.
#
# Gustaf Neumann, Jan 2015
#
package require nsf

#
# Get configured URLs
#
set URLs [ns_config ns/server/[ns_info server]/module/websocket/log-view urls]

if {$URLs eq ""} {
    ns_log notice "WebSocket: no log-view configured ([info script])"
    return
}

#
# Register WebSocket log-viewer under every configured URL
#
foreach url $URLs {
    ns_log notice "WebSocket: log-viewer available under $url"
    ns_register_adp  GET $url [file dirname [info script]]/log-view.adp
    ns_register_proc GET $url/connect ::ws::log::connect
    ns_register_proc GET $url/set ::ws::log::set_logging
}

namespace eval ws::log {
    #
    # The proc "connect" is called, whenever a new WebSocket is
    # established.  Per default, the log-viewer logs the "access.log" of
    # the server. The query parameter "log" can be used to specify
    # other files to be logged, such as e.g. the "error.log". The
    # logfilename is used as channel names for broadcasting to
    # multiple potential subscribers.
    #
    nsf::proc connect {} {
        set log [ns_queryget log "access.log"]
        set channel [ws::handshake -readevent true -callback [list ws::log::send_to_all -log $log]]
        ::ws::subscribe $channel $log
    }

    nsf::proc send_to_all {{-log "access.log"} channel msg} {
        ::ws::multicast $log [ws::build_msg $msg]
    }

    nsf::proc set_logging {} {
        foreach s [ns_logctl severities] {
            set level [ns_queryget $s ""]
            ns_logctl severity $s [expr {$level eq "" ? false : true}]
        }
        ns_return 200 text/plain ""
    }


    nsv_set watch files [list [ns_accesslog file] [ns_info log]]

    ns_thread begindetached -name tail {
        #
        # The global state of the currently watched files is
        # maintained in the associative array "watched". The keys are
        # the filenames, the values are file handles.
        #
        array set ::watched {}

        ns_log notice DETACHED=[ns_info server]

        #
        # Close all files which are currently nor watched
        #
        nsf::proc ::ws::log::close_unwatched {watch_files} {
            foreach fn [array names ::watched] {
                if {$fn ni $watch_files} {
                    ns_log notice "watch: stop watching $fn"
                    close $::watched($fn)
                    unset ::watched($fn)
                }
            }
        }

        #
        # Open all files which are currently watched
        #
        nsf::proc ::ws::log::open_watched {watch_files} {
            foreach fn $watch_files {
                if {![info exists ::watched($fn)]} {
                    ns_log notice "watch: start watching $fn"
                    set ::watched($fn) [open $fn]
                }
            }
        }

        #
        # For every watched file, report changes via multicast
        #
        nsf::proc ::ws::log::report_updates {} {
            foreach {filename f} [array get ::watched] {
                set size [file size $filename]
                set pos  [tell $f]
                if {$pos == 0} {
                    #
                    # first, step to the end and start to watch from there
                    #
                    seek $f 0 end
                } elseif {$size < $pos} {
                    #
                    # file shrunk, restart
                    #
                    ns_log notice "tail: file $filename shrunk"
                    close $::watched($filename)
                    unset ::watched($filename)
                } elseif {$size > $pos} {
                    #
                    # report delta
                    #
                    set delta [read $f]
                    #ns_log notice "tail: multicast <$delta>"
                    ws::multicast $filename [ws::build_msg $delta]
                }
            }
        }

        #
        # For every watched file, report changes via multicast
        #
        if {[catch {
            while {1} {
                set watch_files [nsv_get watch files]
                ws::log::close_unwatched $watch_files
                ws::log::open_watched $watch_files
                ws::log::report_updates
                after [ns_config ns/server/[ns_info server]/module/websocket/log-view refresh 1000]

            }
        } errorMsg]} {
            ns_log warning "file watcher returned $errorMsg"
        }
    }
}

#
# Local variables:
#    mode: tcl
#    tcl-indent-level: 4
#    indent-tabs-mode: nil
# End:


websocket-procs.tcl
#
# WebSocket implementation based on NaviServer's ns_connchan with
# internal buffering.
#
package require nsf
ns_logctl severity Debug(ws) false

#
# For more intense debugging, you might activate (some) the following
# debug levels.
#

#ns_logctl severity Debug(ws) true
#ns_logctl severity Debug(connchan) true

#
# Activate/Deactivate a simple WebSocket echo service (e.g. for testing)
#
namespace eval ::ws {}
set ::ws::echo_service 1

#
# Make sure that we have a sufficiently recent version of
# NaviServer. Otherwise bail out with an exception.
#
catch {ns_connchan read "" ""} errorMsg
if {![string match "*-websocket*" $errorMsg]} {
    error "Please upgrade to a newer version of NaviServer"
}

namespace eval ::ws {

    nsf::proc ::ws::log {args} {
        #
        # Support single and multi argument log messages (like
        # "ns_log").
        #
        ns_log Debug(ws) {*}$args
    }

    nsf::proc ::ws::handshake {
        {-readevent:boolean true}
        {-callback ""}
    } {
        #::ws::log handshake

        set h [ns_conn headers]
        if {[ns_set iget $h upgrade] eq "websocket"} {
            set key               [ns_set iget $h Sec-WebSocket-Key]
            set client_protocols  [ns_set iget $h Sec-WebSocket-Protocol]
            ::ws::log "key: $key Client Protocols: '$client_protocols'"

            if {[llength $client_protocols] > 0} {
                set protocol_line "\r\nSec-WebSocket-Protocol: [lindex $client_protocols 0]"
            } else {
                set protocol_line ""
            }

            set guid "258EAFA5-E914-47DA-95CA-C5AB0DC85B11"
            set reply [ns_crypto::md string -digest sha1 -encoding base64 $key$guid]

            #
            # Make sure, to send the upgrade command in a single sweep
            # (single send buffer), no matter how the server is
            # configured.  Otherwise, we could run into an issue with
            # the current revproxy.
            #
            # In Tcl 8.6, we should use [string cat ...] instead of
            # the "append" stunt.
            set _ {}
            ns_write [append _ \
                          "HTTP/1.1 101 Switching Protocols\r\n" \
                          "Upgrade: websocket\r\n" \
                          "Connection: Upgrade\r\n" \
                          "Sec-WebSocket-Accept: ${reply}${protocol_line}\r\n\r\n"]
            #
            # Unplug the connection channel from the current connection
            # thread. The currently unplugged channels can be queried via
            # "ns_connchan list"
            #
            set handle [ns_connchan detach]

            # event based reading
            if {$readevent} {
                # register a callback to be executed when the channel is readable
                ns_connchan callback $handle [list ws::io $handle $callback] rex
            }

            return $handle

        } else {
            ns_write "HTTP/1.0 200 OK\r\nContent-type: text/plain\r\n\r\n"
            return ""
        }
    }

    nsf::proc ::ws::receive_msg {channel} {
        while {1} {

            try {
                ns_connchan read -websocket $channel
            } trap {POSIX ECONNRESET} {} {
                #
                # The other side has closed the connection. Don't
                # complain and perform standard cleanup.
                #
                ns_log warning "ws::receive_msg: peer has reset connection on $channel"
                dict set d continue 0
                break

            } on error {errorMsg} {
                ns_log error "ws::receive_msg error on $channel: $errorMsg ($::errorCode)"
                dict set d continue 0
                break

            } on ok {d} {
            }
            #ns_log notice "GOT <$d>"

            if {[dict exists $d payload]} {
                #
                # The dictionary contains "payload" only when also the
                # "fin" bit of the frame was set. Since we got some
                # data, receive_msg can report it back.
                #
                if {[dict get $d opcode] == 1} {
                    dict set d payload [encoding convertfrom utf-8 [dict get $d payload]]
                }
                dict set d continue 1
                break

            } elseif {[dict get $d bytes] < 0} {
                #
                # There must have been an error condition.
                #
                ::ws::log "ws::receive_msg on $channel got [dict get $d bytes] bytes"
                dict set d continue 0
                break

            } elseif {[dict get $d bytes] == 0 && ![dict get $d havedata]} {
                #
                # We got no fresh data, but maybe we have still some
                # more unprocessed either in the frame buffer or in
                # the segments buffer.
                #
                ::ws::log "ws::receive_msg on $channel: check unprocessed $d"
                set unprocessed [dict get $d unprocessed]
                incr unprocessed [dict get $d fragments]
                dict set d continue [expr {$unprocessed > 0}]
                ns_log warning "ws::receive_msg on $channel should stop?" \
                    " $d unprocessed $unprocessed -> [dict get $d continue]"

            } elseif {[dict exists $d fin]
                      && [dict get $d fin] == 0
                      && [dict get $d frame] eq "complete"
                  } {
                #
                # The frame is complete, but not final, we have to
                # continue.
                #
                dict set d continue 1

            } elseif {[dict get $d bytes] > 0
                      && [dict get $d frame] eq "incomplete"
                      && [dict get $d unprocessed] > 0
                      && ![dict get $d havedata]
                  } {
                #
                # We got some data, but the data is not sufficient
                # to process the frame.  So, we need more data to
                # fill up this frame.
                #
                ns_log warning "ws::receive_msg need more data <$d>"
                dict set d continue 1

            } else {
                #
                # There might be potentially more cases requiring
                # special handling, but for the time being, we
                # continue.
                #
                ns_log warning "ws::receive_msg essentially unhandled case <$d>"
                dict set d continue 1
            }

            if {![dict get $d havedata]} {
                break
            }
        }
        ::ws::log "ws::revceive_msg returns $d"
        return $d
    }

    #
    # ::ws::io is the callback for readable or writable conditions
    # (and for handling timeouts or errors). The handlers switching
    # between readable and writable conditions whenever an output
    # channel is saturated. In readmode (default) it reads data, when
    # the channel is readable, in write mode the handler is delivering
    # data on the channel whenever it is writable. This is necessary
    # to handle partial write operations in an event driven fashion.
    #

    nsf::proc ::ws::io {
        channel
        callback
        when
    } {
        ::ws::log "ws::io channel $channel callback $callback when $when"
        switch $when {
            r {
                return [::ws::io_readable $channel $callback $when]
            }
            w {
                return [::ws::io_writable $channel $callback $when]
            }
            default {
                ::ws::log "ws::io channel $channel received when <$when>"
                return 0
            }
        }
    }

    nsf::proc ::ws::io_readable {
        channel
        callback
        when
    } {
        #
        # When continue is 1, the event will fire again; when continue
        # is 0 channel will be closed.  A continue of 2 means cancel
        # the callback, but don't close the channel.  Per default, set
        # continue to 1 and in terminating cases to 0.
        #
        set continue 1

        #
        # One physical read operation from the channel might contain
        # multiple messages. In these cases, the dict member
        # "havedata" is set, and we can try to get further messages
        # from the receive buffer contained in the connection channel.
        #
        while {$continue} {
            set d [::ws::receive_msg $channel]
            set continue [dict get $d continue]
            if {$continue && [dict exists $d payload]} {
                switch [dict get $d opcode] {
                    1 -
                    2 { if {$callback ne ""} { {*}$callback $channel [dict get $d payload] }}
                    8 { set continue 0 }
                    9 { ws::send $channel [ns_connchan wsencode -opcode pong "PONG"] }
                    default { }
                }
            }
            if {[dict exists $d havedata] && ![dict get $d havedata]} {
                break
            }
        }

        log "ws::io_readable returns $continue (channel $channel)"
        return $continue
    }

    nsf::proc ::ws::io_writable {
        channel
        callback
        condition
    } {
        # When continue is 1, the event will fire again; when continue
        # is 0 channel will be closed.  A continue of 2 means cancel
        # the callback, but don't close the channel.
        #
        log "ws::io_writable on $channel (condition $condition)"

        set result [ns_connchan write -buffered $channel ""]
        set status [ns_connchan status $channel]
        log "ws::io_writable result <$result> status $status"
        if {$result == 0 || [dict get $status sendbuffer] > 0} {
            ns_log warning "ws::io_writable was not successful flushing the buffer " \
                "(still [dict get $status sendbuffer])... trigger again. status: $status"
            set continue 1
        } else {
            #
            # All was sent, fall back to normal read-event driven handler
            #
            set continue 1
            #ns_log notice "ws::io_writable all was sent, register callback for reading "
            ns_connchan callback $channel [list ws::io $channel $callback] rex
            #ns_log notice "ws::io_writable all was sent, register callback for reading DONE"
        }

        log "ws::io_writable returns $continue (channel $channel)"
        return $continue
    }

    #
    # Subscribe to a named WebSocket channel feed.
    #
    nsf::proc ::ws::subscribe {
        channel
        subscription
    } {

        ns_log notice "ws::subscribe $channel $subscription"

        ns_mutex eval [nsv_get ws subscription_mutex] {
            set subscribers [nsv_lappend ws multicast-$subscription $channel]
        }

        ::ws::log "subscribers of $subscription: $subscribers"
        return $subscribers
    }

    #
    # Cancel a subscription to a named WebSocket feed
    #
    nsf::proc ::ws::unsubscribe {
        channel
        subscription
    } {
        ns_log notice "unsubscribing $channel"

        if {[nsv_get ws "multicast-$subscription" subscribers]} {
            ns_mutex eval [nsv_get ws subscription_mutex] {
                set idx [lsearch -exact $subscribers $channel]
                nsv_set ws "multicast-$subscription" [lreplace $subscribers $idx $idx]
            }
        }
    }

    #
    # Send a WebSocket message to "all" or to subscribers of a named
    # feed. It is expected that the msg is already already encoded as
    # a frame.
    #
    nsf::proc ::ws::multicast {
        {-exclude ""}
        subscription
        msg
    } {
        if {$subscription ne "all"} {
            #
            # Send message to all subscribers
            #
            #::ws::log "ws::multicast send to subscriber of $subscription"
            if {[nsv_get ws "multicast-$subscription" channels]} {
                foreach channel $channels {
                    #::ws::log "Sending to $channel"
                    if {$channel ni $exclude} {
                        if {![ws::send $channel $msg]} {
                            # we got an error, the channel is probably closed
                            ws::log "ws::multicast: automatically unsubscribe $channel from $subscription due to error"
                            ws::unsubscribe $channel $subscription
                        }
                    }
                }
            }
        } else {
            #
            # Send message to all shared channels. This is very
            # dangerous, since probably connchans might be used also
            # for other applications.
            #
            #::ws::log "send to all <[ns_connchan list]> except <$exclude>"
            foreach channel_info [ns_connchan list] {
                lassign $channel_info channel
                if {$channel ni $exclude} {
                    ws::log "Sending to $channel"
                    ws::send $channel $msg
                }
            }
        }
    }

    #
    # Send a WebSocket frame (built by ns_connchan wsencode) to a
    # single client.
    #
    # A result of 1 means success, 0 means something went wrong,
    # connection should be terminated.
    #
    nsf::proc ::ws::send {
        channel
        msg
    } {
        #::ws::log "ws::send $channel"

        try {
            ns_connchan write -buffered $channel $msg

        } on ok {nrBytesSent} {
            #ns_log notice "ws::send $channel -> $nrBytesSent"
            set toSend [string length $msg]
            if {$nrBytesSent < $toSend} {
                set status [ns_connchan status $channel]
                ns_log warning "ws::send: partial write on $channel: " \
                    "actually sent $nrBytesSent toSend $toSend\n$status"
                #
                # We must register a callback for the socket becoming writable again.
                #
                set previousCallback [dict get $status callback]
                set previousCondition [dict get $status condition]
                if {$previousCondition ne "wex"} {
                    set callback [lindex $previousCallback 2]
                    ns_connchan callback $channel [list ws::io $channel $callback] wex
                    ::ws::log "Switch registered callback <$callback> wex"
                }
            }
            return 1
        } on error {errorMsg} {
            ns_log warning "ws::send $channel returned error: $errorMsg"
            return 0
        }
    }

    #
    # Initialize package.
    #
    if {![nsv_exists ws mutex]} {
        nsv_set ws subscription_mutex [ns_mutex create websocket-subscription]
    }
}

namespace eval ::ws::client {
    #
    # Simple WebSocket client implementation
    #
    nsf::proc ::ws::client::open {url} {
        set d [ns_parseurl $url]
        set host [dict get $d host]
        set proto [expr {[dict get $d proto] eq "ws"
                         ? "http"
                         : [dict get $d proto] eq "wss"
                         ? "https"
                         : "unknown"}]
        if {$proto eq "unknown"} {
            error "protocol must be ws:// or wss://"
        }
        set location ${proto}://$host
        if {[dict exists $d port]} {
            append location :[dict get $d port]
            append host :[dict get $d port]
        }
        set request_url $location
        if {[dict get $d path] ne ""} {
            append request_url /[dict get $d path]
        }
        if {[dict get $d tail] ne ""} {
            append request_url /[dict get $d tail]
        }

        set nonce [ns_crypto::randombytes -encoding base64 16]
        set headers [ns_set create headers \
                         Host $host \
                         Upgrade websocket \
                         Connection Upgrade \
                         Cache-Control no-cache \
                         Origin $location \
                         Sec-WebSocket-Key $nonce \
                         Sec-WebSocket-Version 13]
        ::ws::log [list ns_connchan open -headers $headers -version 1.1 -hostname $host $request_url]
        set chan [ns_connchan open -headers $headers -version 1.1 -hostname $host $request_url]
        set replyMsg [ns_connchan read $chan]

        set firstline 1
        set reply [ns_set create reply]

        foreach line [split $replyMsg \n] {
            set line [string trimright $line]
            if {$line eq ""} continue
            if {$firstline} {
                set firstline 0
                ns_set put $reply :status [lindex $line 1]
                continue
            }
            #ns_log notice "<$line>"
            ns_parseheader $reply $line
        }
        if {[ns_set get $reply :status] ne 101} {
            ns_log Warning "reply: [ns_set array $reply]"
            error "ws::client::open returned unexpected status code [ns_set get $reply :status]"
        }
        return $chan
    }

    nsf::proc ::ws::client::send {chan msg} {
        ns_connchan write $chan [ns_connchan wsencode -mask $msg]
    }

    nsf::proc ::ws::client::receive {chan} {
        #
        # Read a single message and return it
        #
        while {1} {
            #
            # In case the received message does not have opcode 1,
            # read on until either sich a message or an error occurs.
            #
            set d [::ws::receive_msg $chan]
            if {[dict exists $d payload]} {
                switch [dict get $d opcode] {
                    1 {
                        #
                        # Here we are done, we received the full
                        # message with the right opcode.
                        #
                        return [dict get $d payload]
                    }
                    default { ns_log notice "no special handling of opcode $opcode"}
                }
            } elseif {[dict get $d frame] eq "exception" || [dict get $d continue] != 1} {
                ns_log Warning "exception on websocket: $d"
                break
            }
        }
    }

    nsf::proc ::ws::client::close {chan} {
        ns_connchan close $chan
    }

    nsf::proc -deprecated ::ws::build_msg {
        {-opcode "text"}
        {-mask:switch}
        payload
    } {
        #
        # This function is just easier migration, since "connchan
        # wsencode" provides a superset of functionality.
        #
        set maskArg [expr {$mask ? "-mask" : ""}]
        return [ns_connchan wsencode {*}$maskArg -opcode $opcode $payload]
    }

}

if {$::ws::echo_service} {
    ns_register_proc GET /websocket/echo ::ws::echo::connect
    namespace eval ws::echo {
        nsf::proc connect {} {
            set chat [ns_conn url]
            set channel [ws::handshake -callback [list ws::echo::send -chat $chat]]
            ws::subscribe $channel $chat
        }

        nsf::proc send {
            {-chat "chat"}
            channel msg
        } {
            #ns_log notice "ws::test::echo call send"
            set r [::ws::send $channel [ns_connchan wsencode $msg]]
            #ns_log notice "ws::test::echo returns <$r>"
            return $r
        }
    }
}

#
# Local variables:
#    mode: tcl
#    tcl-indent-level: 4
#    indent-tabs-mode: nil
# End:
